samples/template/src/main.c

```diff
@@ -0,0 +1,517 @@
+		break;
+	case APP_EVENT_ERROR:
+		LOG_ERR("sid state err %d", (int)sid_get_error(ctx->handle));
+		break;
+	case APP_EVENT_LINK_SWITCH:
+		sid_error_t e = SID_ERROR_NONE;
+
+		static uint32_t link_mask = SID_LINK_TYPE_1;
+
+		switch (link_mask) {
+		case SID_LINK_TYPE_1:
+			link_mask = SID_LINK_TYPE_2;
+			break;
+		case SID_LINK_TYPE_2:
+			link_mask = SID_LINK_TYPE_3;
+			break;
+		default:
+			link_mask = SID_LINK_TYPE_1;
+			break;
+		}
+
+		LOG_INF("Sidewalk link switch to 0x%x", link_mask);
+		ctx->config.link_mask = link_mask;
+
+		if (ctx->handle != NULL) {
+			e = sid_deinit(ctx->handle);
+			if (e) {
+				LOG_ERR("sid deinit err %d", (int)e);
+			}
+		}
+
+		e = sid_init(&ctx->config, &ctx->handle);
+		if (e) {
+			LOG_ERR("sid init err %d", (int)e);
+		}
+
+		e = sid_start(ctx->handle, link_mask);
+		if (e) {
+			LOG_ERR("sid start err %d", (int)e);
+		}
+		break;
+	case APP_EVENT_READY:
+	case APP_EVENT_NOT_READY:
+		break;
+	default:
+		LOG_WRN("event: unknow %d", ctx->sm.event);
+		break;
+	}
+}
+
+static void state_init_entry(void *o)
+{
+	LOG_INF("state: init (entry)");
+
+	struct sid_ctx_s *ctx = CONTAINER_OF(o, struct sid_ctx_s, sm);
+	sid_error_t e = SID_ERROR_NONE;
+
+	e = application_pal_init();
+	if (e) {
+		LOG_ERR("pal init err %d", (int)e);
+	}
+	e = sid_init(&ctx->config, &ctx->handle);
+	if (e) {
+		LOG_ERR("sid init err %d", (int)e);
+	}
+	e = sid_start(ctx->handle, SID_LINK_TYPE_1);
+	if (e) {
+		LOG_ERR("sid start err %d", (int)e);
+	}
+}
+
+static void state_init_run(void *o)
+{
+	LOG_INF("state: init");
+
+	struct sid_ctx_s *ctx = CONTAINER_OF(o, struct sid_ctx_s, sm);
+
+	switch (ctx->sm.event) {
+	case APP_EVENT_SIDEWALK:
+	case APP_EVENT_ERROR:
+		break;
+	case APP_EVENT_READY:
+		smf_set_state(SMF_CTX(&ctx->sm), &app_states[STATE_SIDEWALK_READY]);
+		break;
+	case APP_EVENT_NOT_READY:
+		smf_set_state(SMF_CTX(&ctx->sm), &app_states[STATE_SIDEWALK_NOT_READY]);
+		break;
+	default:
+		LOG_WRN("event: unknow %d", ctx->sm.event);
+		break;
+	}
+}
+
+static void state_not_ready_run(void *o)
+{
+	LOG_INF("state: not ready");
+
+	struct sid_ctx_s *ctx = CONTAINER_OF(o, struct sid_ctx_s, sm);
+
+	switch (ctx->sm.event) {
+	case APP_EVENT_SIDEWALK:
+	case APP_EVENT_NOT_READY:
+	case APP_EVENT_ERROR:
+		break;
+	case APP_EVENT_READY:
+		smf_set_state(SMF_CTX(&ctx->sm), &app_states[STATE_SIDEWALK_READY]);
+		break;
+	case APP_EVENT_SEND:
+		smf_set_state(SMF_CTX(&ctx->sm), &app_states[STATE_SIDEWALK_CONNECTING]);
+		break;
+	default:
+		LOG_WRN("event: unknow %d", ctx->sm.event);
+		break;
+	}
+}
+
+static void state_connecting_entry(void *o)
+{
+	LOG_INF("state: connecting (entry)");
+
+	struct sid_ctx_s *ctx = CONTAINER_OF(o, struct sid_ctx_s, sm);
+	sid_error_t e = SID_ERROR_NONE;
+
+	e = sid_ble_bcn_connection_request(ctx->handle, true);
+	if (e) {
+		LOG_ERR("conn req err %d", (int)e);
+		smf_set_state(SMF_CTX(&ctx->sm), &app_states[STATE_SIDEWALK_NOT_READY]);
+	}
+}
+
+static void state_connecting_run(void *o)
```

Comment:
hmm, maybe naming of this state is misleading, Initially I thought we always go through this state, and I was puzzled how it knows it is time to send a message, if in entry callback it just placed the connection request, 

But the graph was very useful, and it explained that it is the only valid path for sending message from not connected state, so we place connection request, and the main state waits for event ready, which again places event to send message, but now we are in correct ready state to do so.

This proved two things: 1. Naming is difficult, 2. Graphs are very useful.

After consulting with the graph, this name is less misleading. We should figure out how to embed the graph into the source code, or make it as easy to access as possible.

Reply:
Good point, I agree this state is a bit tricky. Maybe we should left connection as a different event like it was in template_ble?